engine/resources/hetzner.go

```diff
@@ -0,0 +1,710 @@
+	return fmt.Errorf("invalid Watch call: HetznerRes requires polling")
+}
+
+// CheckApply checks the resource state and applies the resource if the bool
+// input is true. It returns error info and if the state check passed or not.
+func (obj *HetznerRes) CheckApply(apply bool) (bool, error) {
+	var checkOk = true
+	ctx := context.TODO()
+	// Request up-to-date server info from the API.
+	if err := obj.getServerUpdate(ctx); err != nil {
+		return false, errwrap.Wrapf(err, "getServerUpdate failed")
+	}
+	// Check if the server instance is in the correct state.
+	if c, err := obj.checkApplyServerState(ctx, apply); err != nil {
+		return false, errwrap.Wrapf(err, "")
+	} else if !c {
+		checkOk = false
+	}
+	// Skip further steps if not in the intended state.
+	if !apply && !checkOk {
+		return false, nil
+	}
+	// Skip further steps if the server is not running.
+	if obj.State != StateRunning {
+		return checkOk, nil
+	}
+	// Changes in cpu, location and/or image require a server rebuild.
+	if c, err := obj.checkApplyServerRebuild(ctx, apply); err != nil {
+		return false, errwrap.Wrapf(err, "checkApplyServerRebuild failed")
+	} else if !c {
+		checkOk = false
+	}
+	// Changes in rescue mode can be made while the server is running.
+	if c, err := obj.checkApplyRescueMode(ctx, apply); err != nil {
+		return false, errwrap.Wrapf(err, "checkApplyRescueMode failed")
+	} else if !c {
+		checkOk = false
+	}
+	// TODO: checkapply SSH keys
+	return checkOk, nil
+}
+
+// checkApplyServerState powers the server on/off, creates a new instance, or
+// deletes the existing one as needed, to ensure the correct state.
+func (obj *HetznerRes) checkApplyServerState(ctx context.Context, apply bool) (bool, error) {
+	if obj.init.Debug {
+		obj.init.Logf("checkApplyServerState(apply: %t)", apply)
+	}
+	var checkOk = true
+	// Make sure the server exists as intended before further checks.
+	// TODO: add create permission
+	if obj.server == nil {
+		if obj.State == StateAbsent {
+			return true, nil
+		}
+		if !apply {
+			return false, nil
+		}
+		checkOk = false
+		obj.init.Logf("requesting server creation")
+		if err := obj.createServer(ctx); err != nil {
+			return false, errwrap.Wrapf(err, "createServer failed")
+		}
+	}
+	// Make sure the server is in a steady state before further changes.
+	// TODO: test if hetzner api accepts queries while in transient state
+	if err := obj.waitForSteadyState(ctx); err != nil {
+		return false, errwrap.Wrapf(err, "waitForSteadyState failed")
+	}
+	if obj.stateConverged() {
+		return checkOk, nil
+	}
+	// Request changes if the server is not in the intended state yet.
+	// TODO: add permissions (poweron, poweroff, delete)
+	checkOk = false
+	if !apply {
+		return false, nil
+	}
+	switch obj.State {
+	case StateRunning:
+		obj.init.Logf("requesting server poweron")
+		if _, _, err := obj.client.Server.Poweron(ctx, obj.server); err != nil {
+			return false, errwrap.Wrapf(err, "client.server.poweron failed")
```

Comment:
do we want `client.server.poweron`? This is for the end user, not a programmer. If it's for programmer wrap it in obj.init.Debug{} and obj.Logf

Reply:
Just to be sure, your comment is about the relevance of the Logf message for the end user? Or about the error? 

For the former case, I feel like it would be relevant to know when my instances are powering on/off.  I gave those messages the same logging relevance/level as server creation and deletion, because they all affect the operational availability of your server instance. But I find it a bit difficult to draw the line sometimes between standard logging and debug logging, so please correct me if you don't agree.

For the latter case, I rewrote the error messages to "failed to power server on" etc to make them a bit more readable.